### PR TITLE
feat(analyzer): range suppressions

### LINKED ISSUE #4569: Analyzer range suppressions
Implementation fo https://github.com/biomejs/biome/discussions/187

TL;DR:

```js

// biome-ignore-start lint/style/useConst: reason
let foo = "bar";
let name = "jack";
// biome-ignore-end lint/style/useConst: reason
let surname = "frost"'; // this will raise a diagnostic
```

### PR DESCRIPTION
**Pull request by ematipico**
## Summary

Closes #4569 

This PR refactors suppressions by moving all kinds of suppressions inside a big type called `Suppressions`.

It keeps the existing logic of the line suppressions and top-level suppressions, and it adds new logic for **range suppressions**.

## Test Plan

Added new test cases

### CODE REVIEW

File: crates/biome_analyze/src/lib.rs [hunk 1]
```diff
@@ -202,50 +200,48 @@ where
...
-                        top_level_suppressions.range.unwrap_or_default()
+                        range
                     )
                 } else {
                     AnalyzerSuppressionDiagnostic::new(
                     category!("suppressions/unused"),
                     suppression.comment_span,
                     "Suppression comment has no effect. Remove the suppression or make sure you are suppressing the correct rule.",
-                )
+                 )
```
Comment (arendjr) [inline crates/biome_analyze/src/lib.rs:244]: Unnecessary space?

File: crates/biome_analyze/src/lib.rs [hunk 2]
```diff
@@ -581,136 +528,26 @@ where
...
-        }
-
-        if is_top_level_suppression && !misplaced {
-            self.top_level_suppressions.expand_range(range);
-            return ControlFlow::Continue(());
-        }
 
-        if !suppress_all && suppressed_rules.is_empty() && suppressed_instances.is_empty() {
-            return ControlFlow::Continue(());
+            has_suppressions |= true;
```
Comment (arendjr) [inline crates/biome_analyze/src/lib.rs:540]:
```suggestion
            has_suppressions = true;
```

(The effect’s the same, but the intention is clearer.)

File: crates/biome_analyze/src/lib.rs [hunk 3]
```diff
@@ -777,32 +605,55 @@ pub struct AnalyzerSuppression<'a> {
...
     pub(crate) ignore_range: Option<TextRange>,
+
+    /// The kind of `biome-ignore` comment used for this suppression
+    pub(crate) variant: AnalyzerSuppressionVariant,
+}
+
+#[derive(Debug, Clone)]
+pub enum AnalyzerSuppressionVariant {
+    /// biome-ignore
+    Classic,
```
Comment (arendjr) [inline crates/biome_analyze/src/lib.rs:616]: How about calling it `Line`? We already call them `LineSuppression` elsewhere.

**Review by arendjr — APPROVED** [review]
Nice work!

Just clarifying, it’s fine if a range is never closed with `biome-ignore-end`, right? I think that’s alright, it gives people an out if they think it’s annoying `biome-ignore-all` must be specified at the start.

**Comment by ematipico** [comment]
> Nice work!
> 
> Just clarifying, it’s fine if a range is never closed with `biome-ignore-end`, right? I think that’s alright, it gives people an out if they think it’s annoying `biome-ignore-all` must be specified at the start.

Yes that's a valid use case!

**Comment by Snailedlt** [comment]
This looks great! Is it released yet? If not, when can we expect it to be released?

**Comment by arendjr** [comment]
It will be part of the 2.0 release, but no ETA yet.

**Comment by ematipico** [comment]
> @arendjr I am not sure if biome follows semver, if so, is this really a breaking change? Would be great to have it sooner, thanks.

@ayZagen you can read about our versioning here: https://biomejs.dev/internals/versioning/

Also, we already have breaking changes piled in our main branch, so even though it's a new feature, the next release will have breaking changes.

**Pull request merged**